### PR TITLE
DISCO-867: Compares original and new values before saving runs

### PR DESCRIPTION
**Pull request by jlajoie**

### CODE REVIEW

File: course_discovery/apps/api/v1/views/course_runs.py [hunk 1]
```diff
@@ -209,29 +209,47 @@ def update(self, request, *args, **kwargs):
...
             official_run.status = CourseRunStatus.Unpublished
             official_run.save()
 
-        # Sending draft=False triggers the review process for unpublished courses
-        draft = request.data.pop('draft', True)  # Don't let draft parameter trickle down
-        if not draft and course_run.status != CourseRunStatus.Published:
-            request.data['status'] = CourseRunStatus.LegalReview
+        if new_status:
+            course_run = serializer.save(status=new_status)
+        else:
```
Comment (staubina) [inline course_discovery/apps/api/v1/views/course_runs.py]: Nit: you could probably not have an else here.  In no way a blocker.

```
if new_status:
  course_run = serializer.save(status=new_status)
course_run = serializer.save()
```

File: course_discovery/apps/course_metadata/data_loaders/analytics_api.py [hunk 2]
```diff
@@ -2,10 +2,9 @@
 import logging
 
 import pytz
-
+from analyticsclient.client import Client
```
Comment (staubina) [inline course_discovery/apps/course_metadata/data_loaders/analytics_api.py]: Was this just iSort?

**Review by staubina — APPROVED** [review]
Overall the changes look good.

I was trying to think of a way to quickly simplify the changes in course_runs.py, but we can iterate on it if we need to in the future.

File: course_discovery/apps/api/v1/views/course_runs.py (hunk 1 above)
Comment (jlajoie) [reply course_discovery/apps/api/v1/views/course_runs.py]: Don't want to run two saves if we don't need it!

File: course_discovery/apps/course_metadata/data_loaders/analytics_api.py (hunk 2 above)
Comment (jlajoie) [reply course_discovery/apps/course_metadata/data_loaders/analytics_api.py]: Yes :( not sure why my local complained about it, I might undo since I think it failed me now

File: course_discovery/apps/api/v1/views/course_runs.py [hunk 3]
```diff
@@ -209,29 +209,47 @@ def update(self, request, *args, **kwargs):
...
+            course_run = serializer.save()
```
Comment (mikix) [inline course_discovery/apps/api/v1/views/course_runs.py]: This construction feels a little odd to me. I see why you don't want to pass status=None though.

What about having a `save_kwargs = {}` field that you then do `save_kwargs['status'] = CourseRunStatus.Unpublished` and `course_run = serializer.save(**save_kwargs)`?

Just a nit, this is fine regardless.

File: course_discovery/apps/api/v1/views/course_runs.py [hunk 4]
```diff
@@ -209,29 +209,47 @@ def update(self, request, *args, **kwargs):
...
+        serializer.is_valid(raise_exception=True)
 
-        # If changes are made after review and before publish,
-        # revert status to unpublished.
-        if course_run.status == CourseRunStatus.Reviewed:
-            request.data['status'] = CourseRunStatus.Unpublished
+        # Only do the following if the object has been modified
+        new_status = None
+        # Sending draft=False triggers the review process for unpublished courses
+        draft = request.data.pop('draft', True)  # Don't let draft parameter trickle down
```
Comment (mikix) [inline course_discovery/apps/api/v1/views/course_runs.py]: If we want to avoid `draft` trickling down, pop it before handing the data to the serializer.

File: course_discovery/apps/api/v1/views/course_runs.py [hunk 5]
```diff
@@ -209,29 +209,47 @@ def update(self, request, *args, **kwargs):
...
+        if not draft and course_run.status != CourseRunStatus.Published:
+            new_status = CourseRunStatus.LegalReview
```
Comment (mikix) [inline course_discovery/apps/api/v1/views/course_runs.py]: You moved this up. If you leave it below the other status changing block, you dont need the `not new_status` check below.

File: course_discovery/apps/api/v1/views/course_runs.py [hunk 6]
```diff
@@ -209,29 +209,47 @@ def update(self, request, *args, **kwargs):
...
+
+        changed = False
+        for key, new_value in serializer.validated_data.items():
+            original_value = getattr(course_run, key, None)
+            # If the values aren't equal (base compare) and they both aren't None (additional compare for [] and None)
+            if new_value != original_value and not (
+                    isinstance(new_value, list) and new_value is None and original_value is None
```
Comment (mikix) [inline course_discovery/apps/api/v1/views/course_runs.py]: This logic is hard to follow. Could an intermediate variable help reading clarity?

Comment (jlajoie) [reply course_discovery/apps/api/v1/views/course_runs.py]: Yeah probably, can do so 👍

File: course_discovery/apps/api/v1/views/course_runs.py (hunk 4 above)
Comment (jlajoie) [reply course_discovery/apps/api/v1/views/course_runs.py]: It doesn't get serialized but this is a good call. (also I didn't write this :( )

File: course_discovery/apps/api/v1/views/course_runs.py (hunk 3 above)
Comment (jlajoie) [reply course_discovery/apps/api/v1/views/course_runs.py]: Big fan, happy to do so 👍

File: course_discovery/apps/api/v1/views/course_runs.py (hunk 4 above)
Comment (mikix) [reply course_discovery/apps/api/v1/views/course_runs.py]: I wrote the pop - and it was just because it is not serialized today. And we don't want to accidentally serialize it later, if someone adds draft to the serializer for reading purposes (we don't want to write it - just intercept it)

File: course_discovery/apps/api/v1/views/course_runs.py (hunk 1 above)
Comment (staubina) [reply course_discovery/apps/api/v1/views/course_runs.py]: Good catch, in my head there was a return.

**Review by mikix — APPROVED** [review]
:+1:
